File: src/test/java/work/lclpnet/ap2/impl/base/VotedGameQueueTest.java
package work.lclpnet.ap2.impl.base;

import org.junit.jupiter.api.Test;
import work.lclpnet.ap2.api.base.MiniGameManager;
import work.lclpnet.ap2.impl.game.TestMiniGame;

import java.util.List;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class VotedGameQueueTest {

    @Test
    void pollNextGame_noGames_throws() {
        MiniGameManager manager = mock();

        when(manager.getGames())
                .thenReturn(Set.of());

        var queue = new VotedGameQueue(manager, List.of(), 5);

        assertThrows(IllegalStateException.class, queue::pollNextGame);
    }

    @Test
    void pollNextGame_votedFewerThanMinimum_orderAsExpected() {
        MiniGameManager manager = mock();

        TestMiniGame gameA = new TestMiniGame();
        TestMiniGame gameB = new TestMiniGame();
        TestMiniGame gameC = new TestMiniGame();

        when(manager.getGames())
                .thenReturn(Set.of(gameA));

        var queue = new VotedGameQueue(manager, List.of(gameC, gameB), 5);

        assertEquals(gameC, queue.pollNextGame());
        assertEquals(gameB, queue.pollNextGame());
        assertEquals(gameA, queue.pollNextGame());
        assertEquals(gameA, queue.pollNextGame());
        assertEquals(gameA, queue.pollNextGame());
        assertEquals(gameA, queue.pollNextGame());
    }

    @Test
    void preview_fewerThanMinimum_filledUpToMinimum() {
        MiniGameManager manager = mock();

        TestMiniGame game = new TestMiniGame();

        when(manager.getGames())
                .thenReturn(Set.of(game));

        var queue = new VotedGameQueue(manager, List.of(), 5);

        assertEquals(List.of(game, game, game, game, game), queue.preview().stream().limit(5).toList());
    }

    @Test
    void preview_votedFewerThanMinimum_filledUpByGameManager() {
        MiniGameManager manager = mock();

        TestMiniGame gameA = new TestMiniGame();
        TestMiniGame gameB = new TestMiniGame();

        when(manager.getGames())
                .thenReturn(Set.of(gameA));

        var queue = new VotedGameQueue(manager, List.of(gameB), 5);

        assertEquals(List.of(gameB, gameA, gameA, gameA, gameA), queue.preview().stream().limit(5).toList());
    }

    @Test
    void shiftGame_otherGames_unmodified() {
        MiniGameManager manager = mock();

        TestMiniGame gameA = new TestMiniGame();
        TestMiniGame gameB = new TestMiniGame();
        TestMiniGame gameC = new TestMiniGame();

        when(manager.getGames())
                .thenReturn(Set.of(gameA));

        var queue = new VotedGameQueue(manager, List.of(gameB), 5);

        queue.shiftGame(gameC);

        assertEquals(List.of(gameC, gameB, gameA, gameA, gameA), queue.preview().stream().limit(5).toList());
    }
}

File: src/lib/src/test/kotlin/work/lclpnet/ap2/impl/base/VotedGameQueueTest.kt
package work.lclpnet.ap2.impl.base

import net.minecraft.SharedConstants
import net.minecraft.server.Bootstrap
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.function.Executable
import work.lclpnet.ap2.game.MiniGame
import work.lclpnet.ap2.impl.game.TestMiniGame
import work.lclpnet.gaco.ds.queue.VoidQueuePersistence
import java.util.List
import java.util.Set

internal class VotedGameQueueTest {
    @Test
    fun new_noGames_throws() {
        Assertions.assertThrows(IllegalArgumentException::class.java) {
            VotedGameQueue(
                setOf(),
                listOf(),
                5,
                VoidQueuePersistence.instance()
            )
        }
    }

    @Test
    fun pollNextGame_votedFewerThanMinimum_orderAsExpected() {
        val gameA = TestMiniGame()
        val gameB = TestMiniGame()
        val gameC = TestMiniGame()

        val queue = VotedGameQueue(
            setOf(gameA),
            listOf(gameC, gameB),
            5,
            VoidQueuePersistence.instance()
        )

        assertEquals(gameC, queue.pollNextGame())
        assertEquals(gameB, queue.pollNextGame())
        assertEquals(gameA, queue.pollNextGame())
        assertEquals(gameA, queue.pollNextGame())
        assertEquals(gameA, queue.pollNextGame())
        assertEquals(gameA, queue.pollNextGame())
    }

    @Test
    fun preview_fewerThanMinimum_filledUpToMinimum() {
        val game = TestMiniGame()

        val queue = VotedGameQueue(
            setOf(game),
            mutableListOf(),
            5,
            VoidQueuePersistence.instance()
        )

        assertEquals(
            listOf(game, game, game, game, game),
            queue.preview().stream()
                .map(GameQueue.Entry::game)
                .limit(5)
                .toList()
        )
    }

    @Test
    fun preview_votedFewerThanMinimum_filledUpByGameManager() {
        val gameA = TestMiniGame()
        val gameB = TestMiniGame()

        val queue = VotedGameQueue(
            setOf(gameA),
            listOf(gameB),
            5,
            VoidQueuePersistence.instance()
        )

        assertEquals(
            listOf(gameB, gameA, gameA, gameA, gameA),
            queue.preview().stream()
                .map(GameQueue.Entry::game)
                .limit(5)
                .toList()
        )
    }

    @Test
    fun shiftGame_otherGames_unmodified() {
        val gameA = TestMiniGame()
        val gameB = TestMiniGame()
        val gameC = TestMiniGame()

        val queue = VotedGameQueue(
            setOf(gameA),
            listOf(gameB),
            5,
            VoidQueuePersistence.instance()
        )

        queue.setNextGame(gameC)

        assertEquals(
            listOf(gameC, gameB, gameA, gameA, gameA),
            queue.preview().stream()
                .map(GameQueue.Entry::game)
                .limit(5)
                .toList()
        )
    }

    companion object {
        @JvmStatic
        @BeforeAll
        fun bootstrap() {
            SharedConstants.tryDetectVersion()
            Bootstrap.bootStrap()
        }
    }
}